src/stream.rs

```diff
@@ -1246,13 +1248,22 @@ impl SendBuf {
 #[derive(Clone, Debug, Default, Eq)]
 pub struct RangeBuf {
     /// The internal buffer holding the data.
-    data: Vec<u8>,
+    ///
+    /// To avoid neeless allocations when a RangeBuf is split, this can be
+    /// shared between multiple RangeBuf objects, and sliced using the `start`
+    /// and `len` values.
+    data: Arc<Vec<u8>>,
+
+    /// The initial offset within the internal buffer.
+    start: usize,
```

Comment:
Explained individually, but I think it's good to have a short description about the relationship among start, pos, len and off, which is sometimes confusing to read.

Reply:
I expanded the `RangeBuf` description with an overviews of these fields, hopefully it doesn't make things more confusing :)